training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+
+A function or operation that operates on a type can automatically work on an array of that type
+by essentially applying the function to every element of the array.
+
+<a id="ex13"></a>
+#### Example 13: Promotion
+Returning to the factorial example, our `proc` can operate on an `int array` even though it's only defined to accept an `int`.
+```chapel
+proc factorial(n: int) {
+  return * reduce [i in 1..#n] i;
+}
+writeln(factorial(5));
+writeln(factorial([1, 2, 3, 4, 5]));
+```
+```console
+120
+1 2 6 24 120
+```
+
+<a id="zip_and_agg"></a>
+## Zippered Iteration and Aggregation
+
+<a id="zippered_iteration"></a>
+### Zippered Iteration
+[Zippered Iteration](https://chapel-lang.org/docs/users-guide/base/zip.html#zippered-iteration) is simultaneously
+iterating over multiple iterables (usually arrays and ranges) with compatible domains.
+
+The syntax:
+```chapel
+forall (v1, v2, v3, ..., vn) in zip(A1, A2, A3, ..., An) {
+  // for loop iteration `j`, `vi` will refer to the `j`th element of `Ai`
+}
+```
+Since all `Ai` have compatible domains, we know all `vi` will be local to one another on any given loop iteration.
+
+<a id="ex14"></a>
+#### Example 14: Simple Zippered Iteration
+Let's start with a simple example involving zippered iteration:
+```Chapel
+var A: [1..5] real;
+for (a, i, j) in zip(A, 1.., [3, 0, 1, 2, 4]) {
+  a = i**j;
+}
+writeln(A);
+```
+```console
+1.0 1.0 3.0 16.0 625.0
+```
+Notice we have an [unbounded range](https://chapel-lang.org/docs/primers/ranges.html#variations-on-basic-ranges), `1..`,
+so the end bound is determined by the size of the iterables.
+Since in this case the other iterables are length 5, `1..` is equivalent to `1..5`.
+
+<a id="ex15"></a>
+#### Example 15: Zippered Iteration in Arkouda
+Now let's look at an example based on [`getLengths`](https://github.com/Bears-R-Us/arkouda/blob/b86adeb843275b0b86553534ede632acef4d15e2/src/SegmentedString.chpl#L388-L406) in `SegmentedString.chpl`.
+
+First we need some context. An Arkouda `Strings` (aka SegString) is an array of variable length strings.
+```python
+>>> s = ak.array([f"{'s'*(3-i)}{i}" for i in range(3)])
+>>> s
+array(['sss0', 'ss1', 's2'])
+>>> type(s)
+arkouda.strings.Strings
+```
+
+On the server, a SegString is made up of two distributed array components:
+- values: `uint(8)`
+  - the flattened array of bytes in all the strings (null byte delimited)
+- offsets: `int`
+  - the start index of each individual string in the `values`
+
+For the sake of simplicity, we'll treat values as a string array. So for our `s` above, these components look something like
+```
+values = ['s', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2', '\x00']
+offsets = [0, 5, 9]
+```
+For `getLengths`, we want to calculate the length of each individual string including the null terminator.
+
+```chapel
+const values: [0..#12] string = ['s', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2', '\x00'];
+const offsets = [0, 5, 9];
+const size = 3;
```

Comment:
What is the role of `size` here?  Is it the number of offsets?  (in which case, writing `const size = offsets.size;` feels more bulletproof and descriptive).

Totally optional note, but note that each `const` doesn't require its own keyword if you'd rather write this as:

```chapel
const values: [0..#12] string = [...],
      offsets = [0, 5, 9],
      size = 3;
```

(or any combination thereof).  Mostly just good to know that every declaration need not get its own keyword for those who get tired of typing it.

Reply:
`size` is the number of strings in the `SegString` which will always be equivalent to `offsets.size`. I'll add a comment for this, thanks!

and the single keyword for all the declarations is really neat! I don't know how i've gotten this far without seeing that, but I'll def be using it now!

Reply:
Rather than a comment, maybe just change to `const size = offsets.size;`?  Makes it bulletproof and self-descriptive.

Reply:
sorry I actually did both, I just didn't make that clear in the reply!